### ISSUE TITLE
Write a spec to implement and support private messages (DMs)

### ISSUE THREAD

**Issue by hach-que** [Gargron/mastodon#223]
We should write a spec to support the sending private messages to other users in a federated network in a secure manner.  As far as I'm currently aware, this functionality doesn't exist in Mastodon, and doesn't existing in the wider federated network.

The basis for the spec so far is something like:

* Makes use of private/public key pairs.
* Each user's account Atom feed supplies a public key.
* When sending a private message to a user, the sender encrypts the message with the user's public key.  Only the instance on which the target user resides can decrypt the message.
* The encrypted message is then posted into same feed / messaging protocol as public messages, but with an additional flag to indicate that it's encrypted.
* Instances and clients hide messages that they can't decrypt (even if they don't hide the message, all users will see is the encrypted content instead of the actual message)
* Implementations of the protocol that have users receiving private messages can decrypt the content for those users and display it in some form of private messaging area of their system

While I have a rough idea of how I'd like to extend the Atom feed to incorporate this information (e.g. what it would like in https://mastodon.social/users/hq.atom), but I don't yet have a clear picture of the Pubhubsubbub side of the things for real time communications.

I'm probably going to have a go at drafting up an XSD that covers the additional flags that would exist in the Atom format to support this functionality.

**Comment by Gargron** [comment]
Salmon already uses private/public keys to verify origin of a message. I advice against "encrypting" (or using the term "encrypt") messages because it would be dangerously misleading since the server is the owner of its users' private keys, so it's not end-to-end private. Also, people need to be aware of the risk that a rouge server might choose to publish private messages it receives (but this should not be a problem if end-users will be vetting who they're talking to anyway, e-mail is kind of the same risk anyway)

What you bring up is super important though: I really really need new Atom elements to demarkate when a feed is private and you're supposed to send a follow request vs. following directly. (Also an element for demarkating sensitive content would be great). Afaik there are no currently existing Atom extensions that do this.

**Comment by TazeTSchnitzel** [comment]
Regarding DMs, I wonder if it wouldn't be simpler and more effective to hand off to an existing open messaging protocol. That seems fitting with the spirit of the existing technology, which brings together several standards.

**Comment by unsmell** [comment]
this would in turn restrict DMs to users in the same instance of mastodon, right?

**Comment by TazeTSchnitzel** [comment]
Not necessarily. If multiple servers can speak that messaging protocol, it can work. I don't know the specifics, but XMPP has its own concept of “federation”.

**Comment by hach-que** [comment]
I meant "encrypt" in a strictly implementation level; not "encrypt" as a concept that would be communicated to users.  The intention of encrypting the message is to prevent instances listening on PubSub from reading the private messages destined for other instances; that is, if I have an instance in the federation, I shouldn't be able to spy on or read the content of DMs that aren't sent to any users on my instance.

There isn't really anything we can do about sending a message to a rogue instance which doesn't honor the privacy of those messages.  Paranoid implementors can always get users to create the private/public key pair so that the instance itself doesn't hold the private key (but this is not intended as a Mastodon feature, this is just something permitted by the spec).

**Comment by hach-que** [comment]
I think the only way that we could do end-to-end privacy is if we had installable mobile apps, or if we generated a public/private key pair in the browser and used local storage to store it.  The issue with the latter is that clearing your browser content will delete the key, rendering all previous messages unreadable.  The other issue is that when you sign in on a new device, that would generate a new key pair, so you would only be able to read messages on that devices from that point onwards; any previous messages will be encrypted with keys that are only available on your other devices.

**Comment by noiob** [comment]
I think iMessage does this in a way that allows multiple clients, when you sign in a new one the others get the key, too (so you send to the recipient and to all your devices). That would work, I guess

**Comment by Thann** [comment]
What about using WebRTC for private instant messages?  The massages would never be stored on any server for complete privacy, and they would be "deleted" when both parties closed their browsers. Only problem is that both parties would need to online at the same time.a the this should be a different feature called "live chat". Ill take a look at implementing it.

**Comment by hoodiek** [comment]
i believe @Gargron has decided against instant messaging, yes? can this issue then be closed?

**Comment by swaldie** [comment]
Waiting on @Gargron but I do believe this is a wontfix for now.

**Comment by Gargron** [comment]
Yeah, while this might get revisited in the far future it's out of scope for now, soz.

**Comment by tiesselune** [comment]
Hi, I've read the thread and I understand why this would be hard to implement, but still, i fear that a LOT of people will misunderstand the point of direct messages and the way it works, which could result in unwanted scandal the day those nudes and other sensitive stuff leak from a poorly protected server or an ill-intentionned instance admin (proprietary services usually have good economic reasons to make sure their user content is secure, but a random instance admin? Maybe less) . And that would probably result in bad advertising for Mastodon in general...

So why not address the problem by implementing a self-destruct option on direct messages (and their media) to reduce the risk of accumulating a lot of this kind of content on unsecure places? The warning for non-mastodon (or out-of-date instances) would still apply for those, warning users about what they do, and letting them choose if they wanna go risqué. 😉

Another easy solution would be to actually add a sentence in this warning message, stating clearly that direct messages are NOT encrypted (so people actually know it) and thus are not secure enough for highly sensitive messaging. Now that most messaging apps actually inform their users that they use end-to-end encryption, that should be enough to raise the necessary warning flags in people's head. 😀

(Anyway, thanks for the great work on Mastodon; it's great. Really)

**Comment by hach-que** [comment]
So the latest post-mortem indicates "Direct messages" were implemented, though I can't see it in any of the release change logs, which leaves me quite comfused. Does that mean this ended up being implemented?

**Comment by obinoby** [comment]
Hi,
I'd really like to be sure mys private posts remains private even when replicated on other instances. 

Just to reply on a previous post about end to end encryption, maybe this can work :
- the client (app or browser) generate a keys pair
- the client encrypt the private key with the user password
- the encrypted private key is stored on the instance user profile along side with the clear public key
- on a new device, the encrypted private key is fetched from the instance and decrypted using the user password
This can make it possible to have not to worry about loosing the private key.
Then when writing a private message it have to be encrypted in a way that every legitimate recipient can decrypt it (maybe by encrypting it with every single recipient public key).
But e2e will reduce the functionalities of mastodon because new followers will not be able to view old posts until the client is back online and is used to encrypt again every old messages.

But I'd like something less secure than that. After all I've got to trust the instance where I created my account. So what if any instance creates key pairs for each of their users (sign with a key found in the DNS zone of the domain of the instance so that other instances can check that the users keys are trusted by the instance administrators) and keeps all the private posts issued by its users not encrypted. Then only when replicating the message to an other instance (witch can not be trusted since it is not the one the user chose) the source instance encrypt the message with the public key of every follower. A new follower will transparently trigger a task on the source instance to encrypt every past posts with his public key and then replicate them on his instance.

What do you think about that?

Best regards

**Comment by nightpool** [comment]
this is a very, very old issue. discussion on making private messages end to end encrypted is happening on a number of other threads
